V1.4, fixes & intro ukit (#7)
seL4/whitepaper: whitepaper.tex

  difficult. A good way to look at this is to think of seL4 as the
  ``assembly language of operating systems'': very primitive.

  No-one in their right mind would write a complex system, say a
  database or web server, in assembly language, we use higher-level
  programming languages for that. These simplify the task by providing
  higher-level constructs and abstractions, at the expense of losing
  some of the power the hardware provides. In almost all cases, that
  is a winning trade-off.

  Similarly, one should not try to build a complex system directly on
  seL4, but should use a higher-level framework that provides more
  appropriate abstractions, at the expense of introducing some policy
  and taming the power of seL4. Specifically, such a framework should allows developers
  to focus on the code that implements the services, ignore hardware
  complexities, and automate
  much of the system integration.

  There are presently three main
  component frameworks for seL4, all open source: The Microkit, CAmkES
  and Genode.

  The \href{https://trustworthy.systems/projects/microkit/}{seL4 Microkit} reduces
  the complex seL4 API to literally a handful of simple abstractions,

>   designed around modules called \emph{protection domains}. It also

Reviewer: ```suggestion
  designed around components called \emph{protection domains}. It also
```

Just to not overload on the wording, I think it should be components which is what we usually use in documentation.

  \href{https://docs.sel4.systems/projects/sel4/status.html}{seL4
    project status page} for details.

  \SSect{The seL4 Microkit}\label{s:microkit}

  The Microkit is a thin abstraction layer over seL4 that hides much
  of seL4's complexities at the cost of imposing the restriction of a
  static system architecture: At system configuration time, all

>   modules and their authorised communication channels are known.

Reviewer: Same thing here regarding "modules"

  \begin{figure}[b]
    \centering
    \includegraphics[width=0.6\textwidth]{microkit}
    \caption{Microkit abstractions.}
    \label{f:microkit}
  \end{figure}

  \autoref{f:microkit} shows these abstractions (yes, that's all!) The
  \emph{protection domain} (PD) is the process abstraction: it
  represents a program in execution in its own, isolated address
  space. Two PDs can be connected by a \emph{communication channel}
  and may also share \emph{memory regions} that are mapped in one or
  more PDs' address spaces. A PD may be a VM (running a guest OS), but

>   other PDs are oblivious to that fact.

Reviewer: I would prefer to say something along that lines of "A VM is similar to a PD in that it is an isolated component with the additional capability of running a guest OS". The way it's presented to users of Microkit it is a different concept even though conceptually it's basically a PD.

  \begin{Chilli}
    Microkit further simplifies programming by imposing an
    event-handling model. Specifically, each PD has two or three entry
    points: \code{init()}, \code{notified()} and, optionally,
    \code{protected()}. These are provided by the programmer and invoked
    by the framework on particular events: \code{init} is called exactly
    once at system initialisation time and before any of the other

>     handlers are called. \code{notified()} is invoked if the PD at the

Reviewer: ```suggestion
    entry points are called. \code{notified()} is invoked if the PD at the
```

    other end of the channel calls the \code{notify()} API on the
    channel, and \code{protected()} is invoked if the PD at the other
    end of a channel calls the \code{ppcall()} API.

    PDs have scheduling parameters, which include a priority.

>     \code{notify()} calls are asynchronous: they never block, and the
>     when the receiver PD's \code{notified()} function is invoked depends

Reviewer: ```suggestion
    \code{notify()} calls are asynchronous: they never block, and
    when the receiver PD's \code{notified()} function is invoked depends
```

    on priorities and whether the two PDs are executing on the same core
    or not. These calls pass no data to the receiver other than the
    identity of the notifier, and as such behave similarly to
    interrupts. In contrast, a \code{ppcall()} behaves like a normal
    function call, except that the code it executes happens to be in a
    different PD. In particular, \code{ppcall()} is synchronous, i.e.\ it blocks
    the caller until the callee returns from its \code{protected()}
    handler.  It can also pass arguments and return a
    result value. To prevent deadlock,  a \code{ppcall()} is only
    allowed from lower to higher priority.

    PDs providing a \code{protected()} handler are referred to as
    server, and the PD invoking \code{ppcall()} is called a
    client. Servers execute on the client's \emph{scheduling context}
    (see \autoref{s:mcs}) and core, meaning that their usage of the CPU

>     resource is properly accounted to the client.

Reviewer: This is only true in the case of a passive PD, which is currently an opt-in attribute.

Author: Yes, but shouldn't we encourage people to always use passive servers? I'm still looking for a convincing use case for active ones

Author: I guess I can add a Chilli paragraph...

Reviewer: Yes I just don't want to have people thinking it's always going to behave like that when it's opt-in.

Author: Adding a comment in parentheses

  system, basically just two of the PDs at the top of
  \autoref{f:microkit} and the channel between them.)

  The CapDL spec is a precise representation of access rights in the
  system, and it is what seL4 enforces. Which means that once the
  system gets into the state described by the CapDL spec, it is
  guaranteed to behave as described by the SDF spec, and
  therefore the architecture-level description is sufficient for
  further reasoning about security properties.

  So we need assurance that the system boots up into the state
  described by the CapDL spec. We achieve this with a second
  automated step: We generate from CapDL the startup code that, as
  soon as seL4 itself has booted, takes control and generates all the
  seL4 objects referenced by the spec, including the ones representing
  PDs, and distributes the \emph{capabilities} (see
  \autoref{s:caps}) that grant access to those objects according to
  the spec. At the end of the execution of this init code, the system is
  provably in the state described by the CapDL spec, and thus in the
  state represented by the SDF spec.

  \code{libmicrokit}, which maps the Microkit abstractions to the seL4 API,

>   also verified, which guarantees that seL4 system calls are invoked correctly.

Reviewer: Do we want to make this claim? The latest released version of Microkit that people are using has diverged from what the proofs originally worked on and have not been caught up yet.

Author: fair point. In fact, both this and the CapDL mapping verification are about as incomplete as the CAmkES proofs. The note immediately following does introduce the caveat. Is that not enough? I can add that they are based on an older version.

Author: How about changing the note to read:

> At the time of writing, the proofs about Microkit and
  CapDL are not yet complete and based on an older version of the
  Microkit.  Completion is planned for the future.

Reviewer: Ah, I misinterpreted the note about the note as just about the Microkit+capDL connection.

Reviewer 2: As far as I understood the proofs, they do not actually show that they invoke seL4 correctly in any case -- for that you'd need a connection to the seL4 spec. Isn't that what Rob is currently working on?

Author: Yes, they use a manually-derived spec. Which is part of being "incomplete". But I guess it's better not to be more conservative, and just say "verified to be implemented correctly"?
